### ISSUE TITLE
Can't use Ansible requirements file that lives in higher-level directory

### ISSUE THREAD

**Issue by tehranian** [neillturner/kitchen-ansible#72]
TL; DR: I think this is a problem in how `kitchen-ansible` is generating the path to the requirements file within the VM, for passing to `ansible galaxy install -r ...`.

Here's the situation I'm trying to work with: We have an Ansible `requirements.yml`file that lives two directory levels above where we keep our custom roles. Ex:

```
$ tree -a
.
├── README.md
├── playbook.yml
├── requirements.yml
└── roles
    ├── dlpx.common
    │   ├── .kitchen.yml
    │   ├── files
    │   ├── tasks
    │   ├── test
... <many more> ...
```

I'm trying to maintain the principles of don't-repeat-yourself (DRY) and re-use that top-level `requirements.yml` file for our Test Kitchen tests, rather than creating redundant `requirements.yml` files co-located with each `.kitchen.yml` file. Unfortunately I cannot figure out how to reference that `requirements.yml` file from `.kitchen.yml` successfully. Ex:

* If I specify `requirements_path: requirements.yml`, I get the following error because `requirements.yml` is not in the same working dir as `.kitchen.yml`:

```
# .kitchen.yml :
provisioner:
  hosts: test-kitchen
  name: ansible_playbook
  additional_copy_path:
    - "../dlpx.git-client"
    - "../../requirements.yml"
  requirements_path: requirements.yml

====
# Error:
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: Failed to complete #converge action: [No such file or directory - requirements.yml]
>>>>>> ----------------------
```

* I can get around that by modifying `requirements_path` to be the relative path to `../../requirements.yml`, but then when Test Kitchen goes to run `ansible-galaxy install` it gets run with `-r /tmp/kitchen/../../requirements.yml`, where `/tmp/kitchen/../../requirements.yml` is not the correct path to the `requirements.yml` file that was copied over:

```
# .kitchen.yml :
provisioner:
  hosts: test-kitchen
  name: ansible_playbook
  additional_copy_path:
    - "../dlpx.git-client"
    - "../../requirements.yml"
  requirements_path: "../../requirements.yml"

====
# Error:
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: SSH exited (1) for command: [sudo -E bash -c '[ -d /etc/ansible ] || mkdir /etc/ansible' && sudo -E cp /tmp/kitchen/ansible.cfg /etc/ansible && sudo -E cp -r /tmp/kitchen/group_vars /etc/ansible/. && sudo -E cp -r /tmp/kitchen/host_vars /etc/ansible/. && ansible-galaxy install --force -p /tmp/kitchen/roles -r /tmp/kitchen/../../requirements.yml]
>>>>>> ----------------------

# The files lives at:  /tmp/kitchen/requirements.yml
vagrant@default-ubuntu-1404:~$ ls -l /tmp/kitchen/requirements.yml
-rw-r--r-- 1 vagrant vagrant 895 Jul 10 17:34 /tmp/kitchen/requirements.yml
```

I think this is a problem in how `kitchen-ansible` is generating the path to the requirements file within the VM, for passing to `ansible galaxy install -r ...`.

**Comment by trinitronx** [comment]
There is more of a philosophical question here about how unit and integration testing should work.  I'd contend that ideally you'd want to follow a more Object Oriented approach when organizing your tests.  Generally the pattern in the Chef community with cookbooks is to include all things within the cookbook repo.  Since community and publicly shared cookbooks are always distributed as their own git repo project, it is important to include and specify any other external dependencies within that repo structure.  If there are test-kitchen integration tests that need to pull in dependencies, the `Berksfile` or `Cheffile` should be at the top-level of that cookbook's repo directory.

Now that I've said that, I realize Ansible and patterns in it's community are less structured and are generally more free-form.  There are generally no standards for how to do this sort of thing in Ansible, because the community is not as mature and hasn't developed these best practices yet.  Since Ansible roles are generally the community-shareable unit of configuration management code (much like Chef cookbooks or puppet modules), I think when designing a test suite for an ansible role the same general principles apply.

That is to say:  if you are testing a unit of code like a role, include the things and test harnesses required for testing that role either inside the role's `test/*` directory, or inside the role's directory structure.

In this way, anyone in the community can simply `git clone` that role and then run `kitchen test`, and the dependencies which are specified and included within the `requirements.yml` or `Ansiblefile` are pulled in automatically.  This is only possible if the `requirements.yml` is included with the role itself.  If it were only included at the top-level Ansible playbook repo, then when this role's project is cloned or imported by another person, the dependencies are all lost and unknown.  This is the reason dependency hell usually starts.

I understand the drive to try to keep things D.R.Y. in your example, however in my mind there is a clear distinction between a role and a full top-level Ansible playbooks repo.  At the Ansible playbooks repo level, you could create a higher-level `test/integration` directory which includes `test-kitchen` test suites for the playbooks which make use of many other roles (whether they're from public git repos, Ansible Galaxy, or custom in-house site-specific ones).  You would then use the top-level `requirements .yml`, or the top-level `roles` directory to satisfy the playbook-under-test's dependencies.

Hope this makes sense, and isn't seeming like a rant or some type of imposed design or structure for no reason.

**Comment by neillturner** [comment]
We have a number of issues where people want to have tests defined in a different way to test-kitchen. Clearly we can't support everything. It seems that there is a ansible standard around 
ansible-spec: 
  https://github.com/volanja/ansible-sample-tdd
so we may try and implement support for this standard. 
If folks come up with enhancements to the code to support different ways of working then we will look at them but won't actively try an implement these test cases.
